### PR TITLE
Add filtering by kind and sorting by name or kind to the Unified Resources view

### PR DESCRIPTION
**Pull request by bl-nero**
Note that the changes don't replicate the UI design perfectly, but they are the best compromise between the speed of delivery and the actual functionality. They currently reuse existing components.

Screenshot:
image

Related issue: #27462

### CODE REVIEW

**Comment by zmb3** [comment]
Can you include a screenshot or description of where we deviated from the design? It's nice to have that in the PR for historical reasons.

File: lib/auth/auth_with_roles.go [hunk 1]
```diff
@@ -1717,18 +1690,8 @@ func (a *ServerWithRoles) ListUnifiedResources(ctx context.Context, req *proto.L
...
-				return unifiedNameCompare(filteredResources[i], filteredResources[j], isDesc)
-			})
-		case types.ResourceSpecType:
-			sort.SliceStable(filteredResources, func(i, j int) bool {
-				return stringCompare(filteredResources[i].GetKind(), filteredResources[j].GetKind(), isDesc)
-			})
-		default:
-			return nil, trace.NotImplemented("sorting by field %q for unified resource %q is not supported", req.SortBy.Field, types.KindUnifiedResource)
+		if err := filteredResources.SortByCustom(req.SortBy); err != nil {
+			return nil, err
```
Comment (avatus) [inline lib/auth/auth_with_roles.go:1694]:
```suggestion
			return nil, trace.Wrap(err, "sorting unified resources")
```

Comment (avatus) [inline lib/auth/auth_with_roles.go:1694]:
```suggestion
			return nil, trace.Wrap(err, "sorting unified resources")
```

File: e
Comment (zmb3) [inline e:1]: Unintentional e update?

File: web/packages/teleport/src/UnifiedResources/FilterPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+function oppositeSort(sort: SortType): SortType {
+  switch (sort.dir) {
+    case 'ASC':
+      return { ...sort, dir: 'DESC' };
+    case 'DESC':
+      return { ...sort, dir: 'ASC' };
+    default:
+      // Will never happen. Of course.
+      return sort;
+  }
```
Comment (avatus) [inline web/packages/teleport/src/UnifiedResources/FilterPanel.tsx:130]: this is fine. we could also just check for `DESC` and make the default the `ASC` case

File: api/types/constants.go [hunk 3]
```diff
@@ -952,6 +952,9 @@ const (
 
 	// ResourceSpecType refers to a resource field named "type".
 	ResourceSpecType = "type"
+
+	// ResourceSpecKind refers to a resource field named "kind".
+	ResourceSpecKind = "kind"
```
Comment (avatus) [inline api/types/constants.go:957]: I don't believe `Kind` is part of a resources's spec. Maybe just change to `ResourceKind`?

File: web/packages/teleport/src/UnifiedResources/FilterPanel.tsx (hunk 2 above)
Comment (bl-nero) [reply web/packages/teleport/src/UnifiedResources/FilterPanel.tsx:130]: I somehow prefer to be a bit more explicit :)

**Review by avatus — APPROVED** [review]
If possible without too much change, I'd like to get the search bar/filter styles in line with the desgn (just taller/thicker and whatever). but if it requires many external file changes, then we can push this in and separate that out once its merged.

**Comment by bl-nero** [comment]
I'd rather keep intact those already reviewed changes that bring us to the
MVP state, and then add more stuff on top of it in separate PRs.

**Review by zmb3 — APPROVED** [review]

**Pull request merged**